ta/pkcs11/src/pkcs11_attributes.c

```diff
@@ -724,24 +724,18 @@ create_attributes_from_template(struct obj_attrs **out, void *template,
 			type = PKCS11_CKK_AES;
 			break;
 		default:
-			EMSG("Unable to define class/type from mechanism");
```

Comment:
Now we lost trap for unsupported mechanism in here?

Reply:
Thanks for catching this. I will add a TEE_Panic over here. The unsupported mechanism check is already done in optee_client. So if we have reached this default condition the something has gone wrong.

Reply:
I checked out PR https://github.com/OP-TEE/optee_client/pull/241 and didn't notice any checks for that.

I believe original code is quite good and gives nice error to caller:
```c
		default:
			EMSG("Unable to define class/type from mechanism");
			rc = PKCS11_CKR_TEMPLATE_INCOMPLETE;
			goto out;
```

In my opinion this is good error handling in this case and doesn't cause any havoc anywhere.

We could probably give different error thou... perhaps `PKCS11_CKR_MECHANISM_INVALID`?

Reply:
serialize_ck_mecha_params() in libckteec/src/serialize_ck.c does the check in optee_client code.

Reply:
Ok. 

I believe the pre-check in `serialize_ck_mecha_params()` is OK.

However -- I don't think you can so easily do `TEE_Panic()`.

`TEE_Panic()` should only be done when there is grave error in within TA code.

Also it is not strictly required that someone uses libckteec for communication. Eg. trust layer should be entry to the TA.

In here we are validating input from REE and we can easily bail out. I do not see this being in grave error category.